File: lessons/compiled_lang/java/src/h00.associative.java
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.stream.Collectors;

// Not public - see a00.output.java for why.
class H00Associative {
    public static void main(String[] args) {
        // create empty map
        Map<String, Integer> ages = new LinkedHashMap<>();
        // insert one element at a time
        ages.put("bob", 34);
        ages.put("ed", 58);
        ages.put("steve", 32);
        ages.put("ralph", 23);
        ages.put("deb", 46);
        ages.put("kate", 19);

        // enumerate and print keys
        System.out.println("Keys (names):  " + String.join(", ", ages.keySet()));
        // enumerate and print values
        String values = ages.values().stream().map(String::valueOf).collect(Collectors.joining(", "));
        System.out.println("Values (ages): " + values);
    }
}

File: lessons/compiled_lang/cpp/src/h00.associative.cpp
#include <iostream>
#include <string>
#include <unordered_map>

int main() {
    // create empty map
    std::unordered_map<std::string, int> ages;
    // insert one element at a time
    ages["bob"] = 34;
    ages["ed"] = 58;
    ages["steve"] = 32;
    ages["ralph"] = 23;
    ages["deb"] = 46;
    ages["kate"] = 19;

    // enumerate and print keys
    std::cout << "Keys (names):  ";
    bool first = true;
    for (const auto& pair : ages) {
        if (!first) std::cout << ", ";
        std::cout << pair.first;
        first = false;
    }
    std::cout << std::endl;

    // enumerate and print values
    std::cout << "Values (ages): ";
    first = true;
    for (const auto& pair : ages) {
        if (!first) std::cout << ", ";
        std::cout << pair.second;
        first = false;
    }
    std::cout << std::endl;

    return 0;
}
